File: src/project-id.mjs
import path from "node:path";
import { execFileSync } from "node:child_process";
import { MemoryError } from "./errors.mjs";

export function assertProjectId(id) {
  const parts = String(id).split("/");
  if (!id || /\s/.test(id) || id.startsWith("/") || parts.some((p) => p === "" || p === "." || p === ".."))
    throw new MemoryError("refused", `malformed project id ${JSON.stringify(id)}`);
  return id;
}

export function projectIdFromOrigin(url) {
  let u = url.trim();

  // Reject filesystem paths
  if (u.startsWith("/") || u.startsWith(".") || u.startsWith("~") || /^[a-z]:/i.test(u)) {
    throw new MemoryError("refused", `malformed project id origin ${JSON.stringify(u)}`);
  }

  // Check if URL has a scheme
  const hasScheme = /^[a-z][a-z0-9+.-]*:\/\//i.test(u);

  let host, p;

  if (hasScheme) {
    // Strip scheme
    u = u.replace(/^[a-z][a-z0-9+.-]*:\/\//i, "");
    // Strip optional user[:pass]@
    u = u.replace(/^[^@/]+@/, "");
    // Split at first /
    const i = u.indexOf("/");
    host = u.slice(0, i);
    p = u.slice(i + 1);
  } else {
    // Try SCP form only if no scheme
    const scp = /^([^@\s]+@)?([^:/\s]+):(.+)$/.exec(u);
    if (scp) {
      host = scp[2];
      p = scp[3];
    } else {
      // No scheme and not SCP form
      throw new MemoryError("refused", `malformed project id origin ${JSON.stringify(url.trim())}`);
    }
  }

  // Normalize host: lowercase and strip port
  host = host.toLowerCase().replace(/:\d+$/, "");
  // Normalize path: strip trailing slashes and .git
  p = p.replace(/\/+$/, "").replace(/\.git$/, "");

  return assertProjectId(`${host}/${p}`);
}

function git(cwd, args) {
  try { return execFileSync("git", args, { cwd, stdio: ["ignore", "pipe", "ignore"], encoding: "utf8" }).trim(); }
  catch { return null; }
}

export function localProjectId(cwd) { return assertProjectId(`local/${path.basename(path.resolve(cwd))}`); }

export function projectIdFor(cwd) {
  const top = git(cwd, ["rev-parse", "--show-toplevel"]);
  if (!top) return localProjectId(cwd);
  const origin = git(top, ["remote", "get-url", "origin"]);
  if (!origin) return localProjectId(top);
  try {
    return projectIdFromOrigin(origin);
  } catch (e) {
    if (e.code === "refused") return localProjectId(top);
    throw e;
  }
}

File: memory-go/projectid.go
package memory

import (
	"encoding/json"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

var (
	originScheme = regexp.MustCompile(`(?i)^[a-z][a-z0-9+.-]*://`)
	originCred   = regexp.MustCompile(`^[^@/]+@`)
	originScp    = regexp.MustCompile(`^([^@\s]+@)?([^:/\s]+):(.+)$`)
	originPort   = regexp.MustCompile(`:\d+$`)
	originTrail  = regexp.MustCompile(`/+$`)
	originDotGit = regexp.MustCompile(`\.git$`)
	originIsPath = regexp.MustCompile(`(?i)^[a-z]:`)
	projectIDWs  = regexp.MustCompile(`\s`)
)

// AssertProjectID refuses an empty id, whitespace, a leading slash and any
// empty, "." or ".." segment.
func AssertProjectID(id string) (string, error) {
	bad := id == "" || projectIDWs.MatchString(id) || strings.HasPrefix(id, "/")
	if !bad {
		for _, p := range strings.Split(id, "/") {
			if p == "" || p == "." || p == ".." {
				bad = true
				break
			}
		}
	}
	if bad {
		q, _ := json.Marshal(id)
		return "", Errorf(CodeRefused, "malformed project id %s", q)
	}
	return id, nil
}

func malformedOrigin(u string) error {
	q, _ := json.Marshal(u)
	return Errorf(CodeRefused, "malformed project id origin %s", q)
}

// ProjectIDFromOrigin turns a git origin into host/owner/repo: scheme,
// credentials, port and .git stripped, host lowercased, scp form accepted,
// filesystem paths refused.
func ProjectIDFromOrigin(url string) (string, error) {
	u := strings.TrimSpace(url)
	if strings.HasPrefix(u, "/") || strings.HasPrefix(u, ".") || strings.HasPrefix(u, "~") || originIsPath.MatchString(u) {
		return "", malformedOrigin(u)
	}
	var host, p string
	if originScheme.MatchString(u) {
		u = originScheme.ReplaceAllString(u, "")
		u = originCred.ReplaceAllString(u, "")
		i := strings.Index(u, "/")
		if i == -1 {
			return "", malformedOrigin(strings.TrimSpace(url))
		}
		host, p = u[:i], u[i+1:]
	} else {
		m := originScp.FindStringSubmatch(u)
		if m == nil {
			return "", malformedOrigin(strings.TrimSpace(url))
		}
		host, p = m[2], m[3]
	}
	host = originPort.ReplaceAllString(strings.ToLower(host), "")
	p = originDotGit.ReplaceAllString(originTrail.ReplaceAllString(p, ""), "")
	return AssertProjectID(host + "/" + p)
}

func gitQuiet(cwd string, args ...string) (string, bool) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "", false
	}
	return strings.TrimSpace(string(out)), true
}

func LocalProjectID(cwd string) (string, error) {
	abs, err := filepath.Abs(cwd)
	if err != nil {
		return "", err
	}
	return AssertProjectID("local/" + filepath.Base(abs))
}

// ProjectIDFor is projectIdFor(cwd): the origin of the enclosing repo, else
// local/<top level basename>, else local/<cwd basename>.
func ProjectIDFor(cwd string) (string, error) {
	top, ok := gitQuiet(cwd, "rev-parse", "--show-toplevel")
	if !ok {
		return LocalProjectID(cwd)
	}
	origin, ok := gitQuiet(top, "remote", "get-url", "origin")
	if !ok {
		return LocalProjectID(top)
	}
	id, err := ProjectIDFromOrigin(origin)
	if err != nil {
		if CodeOf(err) == CodeRefused {
			return LocalProjectID(top)
		}
		return "", err
	}
	return id, nil
}
